plugins/auto-sizes/hooks.php

```diff
@@ -78,6 +78,25 @@ function auto_sizes_update_content_img_tag( $html ): string {
 }
 add_filter( 'wp_content_img_tag', 'auto_sizes_update_content_img_tag' );
 
+/**
+ * Checks whether the given 'sizes' attribute includes the 'auto' keyword as the first item in the list.
+ *
+ * Per the HTML spec, if present it must be the first entry.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $sizes_attr The 'sizes' attribute value.
+ * @return bool True if the 'auto' keyword is present, false otherwise.
+ */
+function auto_sizes_attribute_includes_valid_auto( string $sizes_attr ): bool {
+	$parts = preg_split( '/\s*,\s*/', strtolower( trim( $sizes_attr ) ) );
+	if ( ! is_array( $parts ) || ! isset( $parts[0] ) ) {
+		return false;
+	}
+
+	return 'auto' === $parts[0];
```

Comment:
Building on @dmsnell's suggestion in https://github.com/WordPress/performance/pull/1445#discussion_r1708845701 but now taking into account we're only looking at the first token in the list, I think this can be further simplified:

```suggestion
	$token = strtok( strtolower( $sizes_attr ), ',' );
	return 'auto' === trim( (string) $token, " \t\f\r\n" );
```

Note that `trim()` without the second arg includes `\0` and `\v` which are not considered whitespace here in the HTML spec.

Reply:
@westonruter Mostly looks good.

> Note that trim() without the second arg includes \0 and \v which are not considered whitespace here in the HTML spec.

I think that's overkill to consider here. There's probably tons of other `trim()` usages in core where, strictly speaking, these characters are not relevant to consider. I think passing the 2nd parameter is unnecessary and would pose the question why it's not done like that elsewhere.

Reply:
to be fair, those tons of other `trim()` usages are part of a body of code that misparses and leads to corruption and worse. I suspect that the reason it's often done like that has more to do with people coding to their familiarity of HTML and not having fitting tools to operate on a semantic level.

why do you think it's overkill to properly parse it when all that involves is adding the second parameter? just curious on this point, because I've come across the idea before that avoiding corruption is overkill, and I don't understand the opposition to it, particularly when it involves almost no more code.

Reply:
(Not responding to Dennis but following up with some other information I learned.)

Aside: I learned some new things about `strtok()`, I was aware that you only had to pass in the string for the first call and that it would remember it for subsequent calls. What I wasn't aware is that it collapses multiple deliminters and skip over the delimiter if it appears at the beginning: https://3v4l.org/eMMYW

See [note](https://www.php.net/manual/en/function.strtok.php#:~:text=argument%20is%20found.-,Note%3A,-This%20function%20behaves) in PHP function docs.

I also [asked Gemini](https://g.co/gemini/share/e9418c9e0dbd) about the history of `strtok()` only needing the `$string` for the first call:

> The unusual behavior of PHP's `strtok()` function, where subsequent calls only require the delimiter, stems from its inheritance from C and performance optimizations. It modifies the original string in-place and uses a static pointer for efficiency. While this offers benefits for large strings or memory-constrained scenarios, it also leads to non-reentrancy, string modification, and limited flexibility. Consider alternatives like `explode()` or `preg_split()` for more flexibility or a non-destructive approach.

Reply:
interesting. thanks @westonruter - I was unaware.

in the version I tossed out there, there are no allocations and no mutation.

Reply:
I think this is a nit-pick discussion. Either approach works and the performance benefit is likely tiny. I'll let myself be convinced by benchmarks comparing the two variants, but I don't feel strongly about updating it at this point.

If you feel strongly about it, feel free to update the PR. :)

Reply:
I've changed my mind on the use of `strtok()` given what @dmsnell found in https://github.com/WordPress/performance/pull/1445#discussion_r1714424178. 

Undone in https://github.com/WordPress/performance/pull/1476.

Reply:
> I think this is a nit-pick discussion. Either approach works 

Yes it's nit-picky in some regard, but also verifiable wrong to claim that either approach works. The code complexity for the correct and incorrect code are roughly on par, but we have clear knowable failure modes for a cascade of leading commas.

It's more like "we don't anticipate seeing the HTML cases where this algorithm leads to broken results and blank images" - that's fine too, that's your call in your own plugin. It just doesn't equivocate with "either approach works" in the same way that using a PCRE function vs. string functions might.